docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,678 @@
+
+```
+**Remarks:**
+- The memory manager will track current permissions for each region, and can
+    determine whether new permissions require an OCall for EMODPR, e.g RW<->RX, RW->R.
+- These APIs should not be used to change EPC page type to PT_TRIM. Trimming pages
+    are done by sgx_mm_uncommit and sgx_mm_dealloc only.
+
+
+### sgx_mm_commit
+
+```
+
+/*
+ * Commit a partial or full range of memory allocated previously with SGX_EMA_COMMIT_ON_DEMAND.
+ * The API will return 0 if all pages in the requested range are successfully committed.
+ * Calling this API on pages already committed has no effect.
+ * @param[in] addr page aligned starting address.
+ * @param[in] length length of the region in bytes of multiples of page size.
+ * @retval 0 The operation was successful.
+ * @retval EINVAL Any requested page is not in any previously allocated regions, or
+ *                 outside the enclave address range.
+ * @retval EFAULT All other errors.
+ */
+int sgx_mm_commit(void *addr, size_t length);
+
+```
+
+### sgx_mm_commit_data
+
+```
+
+/*
+ * Load data into target pages within a region previously allocated by sgx_mm_alloc.
+ * This can be called to load data and set target permissions at the same time,
+ * e.g., dynamic code loading. The caller has verified data to be trusted and expected
+ * to be loaded to the target address range. Calling this API on pages already committed
+ * will fail.
+ * @param[in] addr page aligned target starting addr.
+ * @param[in] length length of data, in bytes of multiples of page size.
+ * @param[in] data data of @length.
+ * @param[in] prot target permission.
+ * @retval 0 The operation was successful.
+ * @retval EINVAL Any page in requested address range is not previously allocated, or
+ *                outside the enclave address range.
+ * @retval EPERM Any page in requested range is previously committed.
+ * @retval EPERM The target permissions are not allowed by OS security policy,
+ *                  e.g., SELinux rules.
+ */
+int sgx_mm_commit_data(void *addr, size_t length, uint8_t *data, int prot);
+
+```
+**Remarks:**
+- The memory manager decides whether OCalls are needed to request OS to make PTE
+permissions changes. No separate sgx_mm_modify call is needed.
+
+Runtime Abstraction Layer
+----------------------------------
+
+To support and use the EMM, an SGX trusted runtime shall implement following
+abstraction layer APIs.
+
+### Exception Handler Registration
+
+```
+/* Return value used by the EMM #PF handler to indicate
+ *  to the dispatcher that it should continue searching for the next handler.
+ */
+#define SGX_EXCEPTION_CONTINUE_SEARCH 0x0
+
+/* Return value used by the EMM #PF handler to indicate
+ *  to the dispatcher that it should stop searching and continue execution.
+ */
+#define SGX_EXCEPTION_CONTINUE_EXECUTION 0xFFFFFFFF
+
+/*
+ * The EMM page fault (#PF) handler
+ *
+ * @param[in] pfinfo info reported in the SSA MISC region for page fault
```

Comment:
```suggestion
 * @param[in] pfinfo info reported in the SSA MISC region for page fault.
```
nit: Be consistent throughout this file in terms of whether param descriptions end in a period or not.